src/message/header/mod.rs

```diff
@@ -14,3 +19,661 @@ mod date;
+        }
+    }
+
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            headers: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub fn get<H: Header>(&self) -> Option<H> {
+        self.get_raw(&H::name()).and_then(|raw| H::parse(raw).ok())
+    }
+
+    pub fn set<H: Header>(&mut self, header: H) {
+        self.set_raw(H::name(), header.display());
+    }
+
+    pub fn remove<H: Header>(&mut self) -> Option<H> {
+        self.remove_raw(&H::name())
+            .and_then(|(_name, raw)| H::parse(&raw).ok())
+    }
+
+    #[inline]
+    pub fn clear(&mut self) {
+        self.headers.clear();
+    }
+
+    pub fn get_raw(&self, name: &str) -> Option<&str> {
+        self.find_header(name).map(|(_name, value)| value)
+    }
+
+    pub fn get_raw_mut(&mut self, name: &str) -> Option<&mut String> {
+        self.find_header_mut(name).map(|(_name, value)| value)
+    }
+
+    pub fn insert_raw(&mut self, name: HeaderName, value: String) {
+        match self.find_header_mut(&name) {
+            Some((_name, prev_value)) => {
+                prev_value.push_str(", ");
+                prev_value.push_str(&value);
+            }
+            None => self.headers.push((name, value)),
+        }
+    }
+
+    pub fn set_raw(&mut self, name: HeaderName, value: String) {
+        match self.find_header_mut(&name) {
+            Some((_, current_value)) => {
+                *current_value = value;
+            }
+            None => {
+                self.headers.push((name, value));
+            }
+        }
+    }
+
+    pub fn remove_raw(&mut self, name: &str) -> Option<(HeaderName, String)> {
+        self.find_header_index(name).map(|i| self.headers.remove(i))
+    }
+
+    fn find_header(&self, name: &str) -> Option<(&HeaderName, &str)> {
+        self.headers
+            .iter()
+            .find(|&(name_, _value)| name.eq_ignore_ascii_case(name_))
+            .map(|t| (&t.0, t.1.as_str()))
+    }
+
+    fn find_header_mut(&mut self, name: &str) -> Option<(&HeaderName, &mut String)> {
+        self.headers
+            .iter_mut()
+            .find(|(name_, _value)| name.eq_ignore_ascii_case(name_))
+            .map(|t| (&t.0, &mut t.1))
+    }
+
+    fn find_header_index(&self, name: &str) -> Option<usize> {
+        self.headers
+            .iter()
+            .enumerate()
+            .find(|&(_i, (name_, _value))| name.eq_ignore_ascii_case(name_))
+            .map(|(i, _)| i)
+    }
+}
+
+impl Display for Headers {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for (name, value) in &self.headers {
+            Display::fmt(name, f)?;
+            f.write_str(": ")?;
+            HeaderValueEncoder::encode(&name, &value, f)?;
+            f.write_str("\r\n")?;
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct HeaderName(Cow<'static, str>);
+
+impl HeaderName {
+    pub fn new_from_ascii(ascii: String) -> Self {
```

Comment:
I'm not sure about the panic here. I guess the rationale is that header names are usually hardcoded and do not come from user input, but it may be better to just `unwrap` our header names and let users decide how to handle errors here.

Reply:
Right. `new_from_ascii_str` is made for hard-coding header names, but `new_from_ascii` is not, so it should definitely have a non panicking API.